File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

var (
	reader = bufio.NewReader(os.Stdin)
	writer = bufio.NewWriter(os.Stdout)
	graph  = [][]int{}
	tree   = []int{}
)

func dfs(start int) {
	for _, v := range graph[start] {
		if tree[v] == 0 {
			tree[v] = start
			dfs(v)
		}
	}
}
func main() {
	defer writer.Flush()
	var n, a, b int
	fmt.Fscanf(reader, "%d\n", &n)
	graph = make([][]int, n+1)
	tree = make([]int, n+1)
	for i := 0; i < n+1; i++ {
		graph[i] = make([]int, 0)
	}
	for i := 0; i < n-1; i++ {
		fmt.Fscanf(reader, "%d %d\n", &a, &b)
		graph[a] = append(graph[a], b)
		graph[b] = append(graph[b], a)
	}
	dfs(1)
	for i := 2; i < n+1; i++ {
		fmt.Fprintf(writer, "%d\n", tree[i])
	}
}

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)
n = int(input())
graph = [[] for i in range(n+1)]

for i in range(n-1):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)

tree = [0 for _ in range(n+1)]
def dfs(start):
    for i in graph[start]:
        if tree[i] == 0:
            tree[i] = start
            dfs(i)
dfs(1)
for x in range(2, n+1):
    print(tree[x])